File: leetcode1800/solution.go
/*
Solution:

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 4.15 MB | 11.54%

- n: length of nums
*/

func maxAscendingSum(nums []int) int {
	current := 0
	res := 0

	var q []int

	for _, num := range nums {
		if len(q) > 0 && q[len(q)-1] >= num {
			q = q[:0]
			current = 0
		}

		q = append(q, num)
		current += num

		if res < current {
			res = current
		}
	}

	return res
}

File: leetcode1800/solution.rs
/*
Solution:

Time: O(n) | Space: O(n)

Runtime: 0 ms | 100.00%
Memory: 2.28 MB | 45.45%

- n: length of nums
*/

impl Solution {
    pub fn max_ascending_sum(nums: Vec<i32>) -> i32 {
        let mut current: i32 = 0;
        let mut res: i32 = 0;

        let mut q: Vec<i32> = Vec::new();

        for &num in nums.iter() {
            if !q.is_empty() && *q.last().unwrap() >= num {
                q.clear();
                current = 0;
            }

            q.push(num);

            current += num;

            res = res.max(current);
        }

        res
    }
}